clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/AutoElasticsearchIndex.scala

```diff
@@ -0,0 +1,61 @@
+package org.broadinstitute.clio.server.dataaccess.elasticsearch
+
+import java.time.OffsetDateTime
+
+import com.sksamuel.elastic4s.http.ElasticDsl._
+import com.sksamuel.elastic4s.mappings.FieldDefinition
+import org.broadinstitute.clio.util.generic.FieldMapper
+
+import scala.reflect.runtime.universe.Type
+
+/**
+  * Builds an ElasticsearchIndex using shapeless and reflection.
+  *
+  * @param indexName The name of the index.
+  * @tparam Document The document being indexed.
+  */
+class AutoElasticsearchIndex[Document: FieldMapper] private[dataaccess] (
+  override val indexName: String
+) extends ElasticsearchIndex[Document] {
+  override def fields: Seq[FieldDefinition] =
+    AutoElasticsearchIndex.getFieldDefinitions[Document]
+}
+
+object AutoElasticsearchIndex {
+  private def getFieldDefinitions[Document: FieldMapper]
+    : Seq[FieldDefinition] = {
+    FieldMapper[Document].fields.toSeq.sortBy {
+      case (name, _) => name
+    } map {
+      case (name, value) => getFieldDefinition(name, value)
+    }
+  }
+
+  private def getFieldDefinition(fieldName: String,
+                                 fieldType: Type): FieldDefinition = {
+    import s_mach.string._
+    val nameSnake = fieldName.toSnakeCase(Lexer.PascalCase)
+    import scala.reflect.runtime.universe.typeOf
+    fieldType match {
+      case tpe if tpe <:< typeOf[Boolean]         => booleanField(nameSnake)
+      case tpe if tpe <:< typeOf[Int]             => intField(nameSnake)
+      case tpe if tpe <:< typeOf[Long]            => longField(nameSnake)
+      case tpe if tpe <:< typeOf[Float]           => floatField(nameSnake)
+      case tpe if tpe <:< typeOf[Double]          => doubleField(nameSnake)
+      case tpe if tpe <:< typeOf[String]          => keywordField(nameSnake)
+      case tpe if tpe <:< typeOf[OffsetDateTime]  => dateField(nameSnake)
```

Comment:
I haven't played with https://github.com/timeoutdigital/docless. But if it's doing what you need for our large case classes, except for `OffsetDateTime`, then in theory as it's based on shapeless one could patch it for this "custom" type.